### PR TITLE
Ram: Steer to 0/7 MPH support

### PR DESCRIPTION
**Pull request by vincentw56**
This will allow all Rams to steer to 0 or 7 depending on the firmware

### CODE REVIEW

**Comment by vincentw56** [comment]
Route: 0f039a62e5afe2d9|2023-05-27--13-34-29--0

File: selfdrive/car/chrysler/carcontroller.py [hunk 1]
```diff
@@ -2,8 +2,8 @@
 from common.realtime import DT_CTRL
 from selfdrive.car import apply_meas_steer_torque_limits
 from selfdrive.car.chrysler.chryslercan import create_lkas_hud, create_lkas_command, create_cruise_buttons
```
Comment (sunnyhaibin) [inline selfdrive/car/chrysler/carcontroller.py:4]:
```suggestion
from cereal import car
from common.realtime import DT_CTRL
from selfdrive.car import apply_meas_steer_torque_limits
from selfdrive.car.chrysler.chryslercan import create_lkas_hud, create_lkas_command, create_cruise_buttons
```

File: selfdrive/car/chrysler/carcontroller.py [hunk 2]
```diff
@@ -2,8 +2,8 @@
...
-from selfdrive.car.chrysler.values import RAM_CARS, CarControllerParams, ChryslerFlags
-
+from selfdrive.car.chrysler.values import RAM_CARS, RAM_DT, RAM_HD, CarControllerParams, ChryslerFlags
+from cereal import car
```
Comment (sunnyhaibin) [inline selfdrive/car/chrysler/carcontroller.py:7]:
```suggestion
from selfdrive.car.chrysler.values import RAM_CARS, RAM_DT, RAM_HD, CarControllerParams, ChryslerFlags


```

File: selfdrive/car/chrysler/carcontroller.py [hunk 3]
```diff
@@ -49,12 +49,17 @@ def update(self, CC, CS, now_nanos):
...
+          lkas_control_bit = True
+        if (self.CP.minEnableSpeed >= 14.5)  and (CS.out.gearShifter != car.CarState.GearShifter.drive) :
+          lkas_control_bit = False
+      elif CS.out.vEgo > self.CP.minSteerSpeed:
         lkas_control_bit = True
       elif self.CP.flags & ChryslerFlags.HIGHER_MIN_STEERING_SPEED:
         if CS.out.vEgo < (self.CP.minSteerSpeed - 3.0):
           lkas_control_bit = False
-      elif self.CP.carFingerprint in RAM_CARS:
+      elif self.CP.carFingerprint in RAM_HD:
```
Comment (sunnyhaibin) [inline selfdrive/car/chrysler/carcontroller.py:62]: Does this have to be `RAM_HD` only?
```suggestion
      elif self.CP.carFingerprint in RAM_CARS:
```

File: selfdrive/car/chrysler/interface.py [hunk 4]
```diff
@@ -62,13 +62,17 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long, docs):
...
+      ret.minEnableSpeed = 14.6
+      if car_fw is not None:
+        for fw in car_fw:
+          if fw.ecu == 'eps' and fw.fwVersion in (b"68273275AF", b"68273275AG", b"68312176AE", b"68312176AG", ):
+            ret.minEnableSpeed = 0.
 
     elif candidate == CAR.RAM_HD:
-      ret.steerActuatorDelay = 0.2
+      stiffnessFactor = 0.35
+      ret.steerActuatorDelay = 0.25
```
Comment (sunnyhaibin) [inline selfdrive/car/chrysler/interface.py:75]:
```suggestion
      ret.steerActuatorDelay = 0.2
```
I would prefer to keep this PR for supporting steering down to 0. Let's have tune-specific changes in another PR.

File: selfdrive/car/chrysler/interface.py [hunk 5]
```diff
@@ -62,13 +62,17 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long, docs):
       ret.wheelbase = 3.88
       ret.steerRatio = 16.3
       ret.mass = 2493. + STD_CARGO_KG
-      ret.minSteerSpeed = 14.5
-      # Older EPS FW allow steer to zero
-      if any(fw.ecu == 'eps' and fw.fwVersion[:4] <= b"6831" for fw in car_fw):
-        ret.minSteerSpeed = 0.
+      CarInterfaceBase.configure_torque_tune(candidate, ret.lateralTuning)
```
Comment (sunnyhaibin) [inline selfdrive/car/chrysler/interface.py:65]:
```suggestion
```
Same as:
- https://github.com/sunnyhaibin/sunnypilot/pull/157/files#r1208245022

File: selfdrive/car/chrysler/interface.py [hunk 6]
```diff
@@ -62,13 +62,17 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long, docs):
...
+      ret.minSteerSpeed = 0.5
+      ret.minEnableSpeed = 14.6
```
Comment (sunnyhaibin) [inline selfdrive/car/chrysler/interface.py:67]:
```suggestion
      ret.minEnableSpeed = 14.5
```
Is the additional 0.1 required?

File: selfdrive/car/chrysler/interface.py [hunk 7]
```diff
@@ -62,13 +62,17 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long, docs):
...
+      if car_fw is not None:
+        for fw in car_fw:
+          if fw.ecu == 'eps' and fw.fwVersion in (b"68273275AF", b"68273275AG", b"68312176AE", b"68312176AG", ):
+            ret.minEnableSpeed = 0.
```
Comment (sunnyhaibin) [inline selfdrive/car/chrysler/interface.py:71]:
```suggestion
      # Certain EPS FW allow steer to zero
      if any(fw.ecu == 'eps' and (fw.fwVersion[:4] <= b"6831" or fw.fwVersion[:4] <= b"6827") for fw in car_fw):
        ret.minEnableSpeed = 0.
```

File: selfdrive/car/chrysler/carcontroller.py [hunk 8]
```diff
@@ -49,12 +49,17 @@ def update(self, CC, CS, now_nanos):
 
       # TODO: can we make this more sane? why is it different for all the cars?
       lkas_control_bit = self.lkas_control_bit_prev
-      if CS.out.vEgo > self.CP.minSteerSpeed:
+      if self.CP.carFingerprint in RAM_DT:
+        if CS.out.vEgo >= self.CP.minEnableSpeed and CS.out.vEgo <= self.CP.minEnableSpeed + 0.5:
+          lkas_control_bit = True
+        if (self.CP.minEnableSpeed >= 14.5)  and (CS.out.gearShifter != car.CarState.GearShifter.drive) :
```
Comment (sunnyhaibin) [inline selfdrive/car/chrysler/carcontroller.py:55]:
```suggestion
        if (self.CP.minEnableSpeed >= 14.5) and (CS.out.gearShifter != 2):
```
Let's use the cereal ordinal here so one less import.

File: selfdrive/car/chrysler/interface.py (hunk 4 above)
Comment (vincentw56) [reply selfdrive/car/chrysler/interface.py]: Made the changes

File: selfdrive/car/chrysler/carcontroller.py [hunk 9]
```diff
@@ -2,8 +2,7 @@
 from common.realtime import DT_CTRL
 from selfdrive.car import apply_meas_steer_torque_limits
 from selfdrive.car.chrysler.chryslercan import create_lkas_hud, create_lkas_command, create_cruise_buttons
-from selfdrive.car.chrysler.values import RAM_CARS, CarControllerParams, ChryslerFlags
-
+from selfdrive.car.chrysler.values import RAM_CARS, RAM_DT, RAM_HD, CarControllerParams, ChryslerFlags
```
Comment (sunnyhaibin) [inline selfdrive/car/chrysler/carcontroller.py:5]:
```suggestion
from selfdrive.car.chrysler.values import RAM_CARS, RAM_DT, RAM_HD, CarControllerParams, ChryslerFlags

```

File: selfdrive/car/chrysler/carcontroller.py [hunk 10]
```diff
@@ -49,12 +49,17 @@ def update(self, CC, CS, now_nanos):
 
       # TODO: can we make this more sane? why is it different for all the cars?
       lkas_control_bit = self.lkas_control_bit_prev
-      if CS.out.vEgo > self.CP.minSteerSpeed:
+      if self.CP.carFingerprint in RAM_DT:
+        if CS.out.vEgo >= self.CP.minEnableSpeed and CS.out.vEgo <= self.CP.minEnableSpeed + 0.5:
+          lkas_control_bit = True
+        if (self.CP.minEnableSpeed >= 14.5)  and (CS.out.gearShifter != 2) :
```
Comment (sunnyhaibin) [inline selfdrive/car/chrysler/carcontroller.py:55]:
```suggestion
        if (self.CP.minEnableSpeed >= 14.5) and (CS.out.gearShifter != 2) :
```

File: selfdrive/car/chrysler/carcontroller.py [hunk 11]
```diff
@@ -49,12 +49,17 @@ def update(self, CC, CS, now_nanos):
...
+          lkas_control_bit = True
+        if (self.CP.minEnableSpeed >= 14.5) and (CS.out.gearShifter != 2) :
+          lkas_control_bit = False
+      elif CS.out.vEgo > self.CP.minSteerSpeed:
         lkas_control_bit = True
       elif self.CP.flags & ChryslerFlags.HIGHER_MIN_STEERING_SPEED:
         if CS.out.vEgo < (self.CP.minSteerSpeed - 3.0):
           lkas_control_bit = False
-      elif self.CP.carFingerprint in RAM_CARS:
+      elif self.CP.carFingerprint in RAM_HD:
```
Comment (sunnyhaibin) [inline selfdrive/car/chrysler/carcontroller.py:62]:
```suggestion
      elif self.CP.carFingerprint in RAM_CARS:
```

File: selfdrive/car/chrysler/interface.py [hunk 12]
```diff
@@ -62,10 +62,10 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long, docs):
       ret.wheelbase = 3.88
       ret.steerRatio = 16.3
       ret.mass = 2493. + STD_CARGO_KG
-      ret.minSteerSpeed = 14.5
-      # Older EPS FW allow steer to zero
-      if any(fw.ecu == 'eps' and fw.fwVersion[:4] <= b"6831" for fw in car_fw):
-        ret.minSteerSpeed = 0.
+      ret.minSteerSpeed = 0.5
+      ret.minEnableSpeed = 14.5
```
Comment (sunnyhaibin) [inline selfdrive/car/chrysler/interface.py:66]:
```suggestion
      ret.minEnableSpeed = 14.5
      # Certain EPS FW allow steer to zero
```

**Comment by sunnyhaibin** [comment]
LGTM. Thanks for the PR @vincentw56 ! 🎉

File: selfdrive/car/chrysler/carcontroller.py [hunk 13]
```diff
@@ -2,7 +2,7 @@
 from common.realtime import DT_CTRL
 from selfdrive.car import apply_meas_steer_torque_limits
 from selfdrive.car.chrysler.chryslercan import create_lkas_hud, create_lkas_command, create_cruise_buttons
-from selfdrive.car.chrysler.values import RAM_CARS, CarControllerParams, ChryslerFlags
+from selfdrive.car.chrysler.values import RAM_CARS, RAM_DT, RAM_HD, CarControllerParams, ChryslerFlags
```
Comment (sunnyhaibin) [inline selfdrive/car/chrysler/carcontroller.py:5]:
```suggestion
from selfdrive.car.chrysler.values import RAM_CARS, RAM_DT, CarControllerParams, ChryslerFlags
```

**Pull request merged**